In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import roc_auc_score

In [2]:
data=pd.read_csv(r"C:\Users\SAKET NANDAN\Documents\current_excelr_lms\logistic_regression_assignment\creditcard.csv")

In [3]:
data.isnull().sum()

Unnamed: 0     0
card           0
reports        0
age            0
income         0
share          0
expenditure    0
owner          0
selfemp        0
dependents     0
months         0
majorcards     0
active         0
dtype: int64

In [4]:
    data.drop(labels='Unnamed: 0', axis=1,inplace=True)

In [5]:
#some of the encoder can not able to create realtion ship with categorical type target , so here i m using custom fuction to make the 
#the target as integer 
#target is labeled as integer then no need to do this transformation 

def find_category_mappings(df, variable):
    return {k: i for i, k in enumerate(df[variable].unique(), 0)}

def integer_encode(data,  variable, ordinal_mapping):

    data[variable] = data[variable].map(ordinal_mapping)


In [6]:
for variable in ['card']:
    mappings = find_category_mappings(data, variable)
    integer_encode(data, variable, mappings)

In [8]:
#data

#### in feture selction analysis i found that with only two features in independent variable i m getting 0.93 as roc_auc value for both train and test without overfitting ,so i m using important columns    'reports', 'share'

In [9]:
data.columns

Index(['card', 'reports', 'age', 'income', 'share', 'expenditure', 'owner',
       'selfemp', 'dependents', 'months', 'majorcards', 'active'],
      dtype='object')

In [10]:
    data.drop(labels=['age','income','expenditure','owner','selfemp','dependents','months','majorcards','active'], axis=1,inplace=True)

In [11]:
data

card  reports     share
0        0        0  0.033270
1        0        0  0.005217
2        0        0  0.004156
3        0        0  0.065214
4        0        0  0.067051
...    ...      ...       ...
1314     0        0  0.002146
1315     1        5  0.000376
1316     0        0  0.026513
1317     0        0  0.008999
1318     0        0  0.111619

[1319 rows x 3 columns]

In [12]:
# Let's divide into train and test set

X_train, X_test, y_train, y_test = train_test_split(
    data.drop(labels='card', axis=1),  # predictors
    data['card'],  # target
    test_size=0.3,
    random_state=0)

X_train.shape, X_test.shape

((923, 2), (396, 2))

In [13]:
# the scaler - for standardisation
from sklearn.preprocessing import StandardScaler

In [16]:
# standardisation: with the StandardScaler from sklearn

# set up the scaler
scaler = StandardScaler()
# fit the scaler to the train set, it will learn the parameters
scaler.fit(X_train)

StandardScaler(copy=True, with_mean=True, with_std=True)

In [17]:
X_train_scaled = scaler.transform(X_train)
X_test_scaled = scaler.transform(X_test)

## performing logistic reggression

In [19]:
def run_logistic(X_train, X_test, y_train, y_test):

    # function to train and test the performance of logistic regression
    logit = LogisticRegression(random_state=44, C=0.01)
    logit.fit(X_train, y_train)

    print('Train set')
    pred = logit.predict_proba(X_train)
    print(
        'Logistic Regression roc-auc: {}'.format(roc_auc_score(y_train, pred[:, 1])))

    print('Test set')
    pred = logit.predict_proba(X_test)
    print(
        'Logistic Regression roc-auc: {}'.format(roc_auc_score(y_test, pred[:, 1])))

In [20]:
run_logistic(X_train_scaled, X_test_scaled, y_train, y_test)


Train set
Logistic Regression roc-auc: 0.9700555445348145
Test set
Logistic Regression roc-auc: 0.9622772239116208


In [21]:
data

card  reports     share
0        0        0  0.033270
1        0        0  0.005217
2        0        0  0.004156
3        0        0  0.065214
4        0        0  0.067051
...    ...      ...       ...
1314     0        0  0.002146
1315     1        5  0.000376
1316     0        0  0.026513
1317     0        0  0.008999
1318     0        0  0.111619

[1319 rows x 3 columns]

### so here with only two featured( reports and share) i m getting roc_auc value 0.97 for train and 0.962 for test which pretty high as well as no overfitting here 